x/mongo/driver/topology/server.go

```diff
@@ -273,9 +273,15 @@ func (s *Server) Connection(ctx context.Context) (driver.Connection, error) {
 }
 
 // ProcessHandshakeError implements SDAM error handling for errors that occur before a connection finishes handshaking.
-func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint64) {
-	// ignore nil or stale error
-	if err == nil || startingGenerationNumber < atomic.LoadUint64(&s.pool.generation) {
+func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint64, serverID *primitive.ObjectID) {
+	// Ignore the error if the server is behind a load balancer but the server ID is unknown. This indicates that the
+	// error happened when dialing the connection or during the MongoDB handshake, so we don't know the server ID to use
+	// for clearing the pool.
+	if err == nil || s.cfg.loadBalanced && serverID == nil {
```

Comment:
Note that there's a distinction between `nil` and `primitive.NilObjectID`. The former is a `nil` pointer and the latter is a non-nil pointer that contains 12 `0` bytes. In LB mode, the former is possible when we get here if we encountered an error before the initial `hello` succeeded. The latter is never possible because 12 0 bytes is not a valid ObjectId (at least not right now because the timestamp field would be non-zero).